plymouth: unstable-2021-10-18 -> unstable-2023-06-05

lilyinstarlight opened the pull request:
###### Description of changes

I'm going to go ahead and open up a full PR for review rather than [leaving the diff buried in comments](https://github.com/NixOS/nixpkgs/pull/217728#discussion_r1136088299)

This migrates the plymouth derivation from the old autotools-based build system to the newer meson-based one

This also updates the plymouth NixOS module to use `runstatedir` instead of hardcoding `/etc` since Plymouth can already pull most things from there and works without the sketchy autotools substitution to rewrite cpp definitions

I've tested this with a matrix of configs for `{scripted stage 1, systemd stage 1} x {plain, luks, nixos-bgrt theme, breeze-bgrt theme}` in a VM and watched for (more or less) flickerless boot

Supersedes and closes #217728 and closes #183024

cc: @Kranzes

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] x86_64-linux -> aarch64-linux (cross)
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

lilyinstarlight committed:
nixos/plymouth: add actual logo defaultText and move existing to example

Fixes #221391

jtojnar commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,102 +1,112 @@
   '';
 
+  # HACK: We want to install configuration files to $out/etc
+  # but GDM should read them from /etc on a NixOS system.
```
GDM → plymouth

jtojnar commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,102 +1,112 @@
+  # at install time but Meson does not support this
+  # so we need to convince it to install all files to a temporary
+  # location using DESTDIR and then move it to proper one in postInstall.
+  DESTDIR = "${placeholder "out"}/dest";
```
Should probably be `env.DESTDIR` now (structured-attrs).

jtojnar commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,102 +1,112 @@
+    substituteInPlace meson.build \
+      --replace "run_command(['scripts/generate-version.sh'], check: true).stdout().strip()" "'${finalAttrs.version}'"
+
+    sed -i '/^install_emptydir/d' src/meson.build
```
Please add comment why this is neeeded.

jtojnar commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,102 +1,112 @@
+    "-Dbackground-end-color-stop=0x000000"
+    "-Drelease-file=/etc/os-release"
+    "-Dtracing=true"
+    "-Dsystemd-integration=true"
```
Many of these are enabled by default with Meson port: https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/a5eda165689864cc9a25ec14fd8c6da458598f42/meson_options.txt

jtojnar commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,102 +1,112 @@
+    libevdev
+    libxkbcommon
+    xorg.xkeyboardconfig
+    libxslt
```
Wass not this in `nativeBuildInputs` for `xsltproc` for DocBook handling?

ElvishJerricco commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,102 +1,120 @@
+        rm -r "$DESTDIR/''${!o}"
+    done
+    # Ensure the DESTDIR is removed.
+    rmdir "$DESTDIR/nix/store" "$DESTDIR/nix" "$DESTDIR"
```
This isn't a big deal, but technically the nix store isn't necessarily stored at `/nix/store`.

lilyinstarlight replied:
I copied this from GDM, but I'll adjust it to `builtins.storeDir`. Might as well

ElvishJerricco commented on pkgs/os-specific/linux/plymouth/default.nix:
```diff
@@ -1,102 +1,120 @@
+  # at install time but Meson does not support this
+  # so we need to convince it to install all files to a temporary
+  # location using DESTDIR and then move it to proper one in postInstall.
+  env.DESTDIR = "${placeholder "out"}/dest";
```
This makes me slightly uneasy. Is it possible to tell plymouth where to read files from with the CLI rather than using `/etc`? Then we could just override the `ExecStart` of the service with something that provides user configured files on the CLI. It'd be nice to just avoid this problem altogether.

lilyinstarlight replied:
There is not a good way, no (it's hard-coded as a compiler define and catted to harcoded strings). This snippet was also copied from GDM and makes Plymouth behave as expected by default (load `/etc` instead of `$out/etc`, like most stuff in nixpkgs). Even if we did override stuff on the CLI (which would require custom patching), is there a specific concern with doing this?

I'm not sure I see why this line specifically has to do with passing stuff to CLI

ElvishJerricco commented on nixos/modules/system/boot/plymouth.nix:
```diff
@@ -184,19 +191,24 @@ in
           done
+          # Fixup references
+          for theme in $out/*/*.plymouth; do
+            sed -i "s,/nix/.*/share/plymouth/themes,$out," "$theme"
```
again, *technically* the store isn't necessarily in `/nix/store`. I also *kind of* wonder if we can allow these files to be in their original locations in initrd rather than patching their locations here with `sed`. If the list of files is dynamic, that's a bit tricky, but still doable with a derivation that creates symlinks to the specific files we need.

lilyinstarlight replied:
Without this bad things happen (I didn't originally have it and just had it symlink in)... Themes expect to be able to depend on each other and do so via relative paths that assume they are installed side-by-side

I will adjust it to use `builtins.storeDir` here too though, because we might as well. I'm pretty sure other things in the plymouth module also expect `/nix` (which is why I didn't just use `storeDir` initially), but I'll have to go double check and fix them too then

lilyinstarlight replied:
Oh I remember what the issue now. It's because of the logo/watermark. You need it in the `ImageDir` of the theme, but it's not distributed with Plymouth and we replace it with the `boot.plymouth.logo` option's value

Therefore the `ImageDir` of the Plymouth `spinner` theme needs to be patched regardless unless we want to mess around with weird derivation assumptions or impure locations

lilyinstarlight committed:
plymouth: unstable-2021-10-18 -> unstable-2023-06-05

Includes switch from autotools-based build system to meson-based build
system.

Co-Authored-By: Ilan Joselevich <personal@ilanjoselevich.com>

lilyinstarlight committed:
nixos/plymouth: use new runtime paths for plugins and themes

SuperSandro2000 commented on nixos/modules/system/boot/plymouth.nix:
```diff
@@ -98,7 +98,8 @@ in
         default = "${nixos-icons}/share/icons/hicolor/48x48/apps/nix-snowflake-white.png";
-        defaultText = literalExpression ''pkgs.fetchurl {
+        defaultText = literalExpression ''"''${nixos-icons}/share/icons/hicolor/48x48/apps/nix-snowflake-white.png"'';
+        example = literalExpression ''pkgs.fetchurl {
           url = "https://nixos.org/logo/nixos-hires.png";
           sha256 = "1ivzgd7iz0i06y36p8m5w48fd8pjqwxhdaavc0pxs7w1g7mcy5si";
         }'';
```
```suggestion
        example = literalExpression ''
          pkgs.fetchurl {
            url = "https://nixos.org/logo/nixos-hires.png";
            sha256 = "1ivzgd7iz0i06y36p8m5w48fd8pjqwxhdaavc0pxs7w1g7mcy5si";
          }
        '';
```
otherwise the doc rendering is off

lilyinstarlight replied:
Fixed in https://github.com/NixOS/nixpkgs/pull/242420

SuperSandro2000 commented on nixos/modules/system/boot/plymouth.nix:
```diff
@@ -225,6 +238,11 @@ in
         plymouth-switch-root.wantedBy = [ "initrd-switch-root.target" ];
       };
+      # Link in runtime files before starting
+      services.plymouth-start.preStart = ''
+        mkdir -p /run/plymouth
+        ln -sf /etc/plymouth/{plymouthd.defaults,themes,plugins} /run/plymouth/
+      '';
```
We can't use tmpfiles here?

ElvishJerricco replied:
@SuperSandro2000 `tmpfiles` runs after `local-fs.target`, which is *far* too late for plymouth.
